conans/client/installer.py

```diff
@@ -47,10 +47,12 @@ def _get_build_folder(self, conanfile, package_layout):
         recipe_build_id = build_id(conanfile)
         pref = package_layout.reference
         if recipe_build_id is not None and pref.package_id != recipe_build_id:
-            package_layout.build_id = recipe_build_id
+            conanfile.output.info(f"build_id() computed {recipe_build_id}")
             # check if we already have a package with the calculated build_id
             recipe_ref = pref.ref
             build_prev = self._cache.get_matching_build_id(recipe_ref, recipe_build_id)
+            if build_prev is None:  # Only store build_id of the first one actually building it
```

Comment:
This is the core fix, some more explanations in https://github.com/conan-io/conan/issues/14537#issuecomment-1689766266